src/test/java/io/appium/java_client/android/AndroidTouchTest.java

```diff
@@ -97,9 +100,10 @@ public void setUp() throws Exception {
 
         TouchAction dragNDrop = new TouchAction(driver)
                 .longPress(longPressOptions()
-                        .withOffset(offset(center1.x, center1.y))
+                        .withPosition(coordinates(center1.x, center1.y))
                         .withDuration(ofSeconds(2)))
-                .moveTo(offset(center2.x, center2.y))
+                .moveTo(position()
+                        .withPosition(coordinates(center2.x, center2.y)))
```

Comment:
how this expression visually helps to understand, that this position is calculated relatively to the previous element and the previous one is absolute?

Reply:
@mykola-mokhnach 
The signature of methods prevents the setting of the wrong offset. Like 
```java
public T tap(Position<PointOption> tapOptions) 
```
It means that user may set only absolute offset (`PointOption`, `coordinates(x, y)`) or element-option there. 

You said before
>moveTo: relative coordinates (x, y) and partially absolute coordinates (element; element, x ,y)

It means that `moveTo` is not bounded by absolute/relative/element offsets formally.  So I desided to 

```java
public <S extends AbstractPositionOption> T moveTo(Position<S> moveToOptions) {
```

and it is possible to

```java
moveTo(position().withOffset(coordinates()))
moveTo(position().withOffset(offset()))
moveTo(position().withOffset(element()))
```

I thought that one of ideas behind these changes was to minimize method count and exclude unnecessary overloading/code repeating. But, yes, it should be safe and it should exclude accidental user's actions.

If not everything is ok with this approach then I think that we could only split `moveTo` into
```java
moveTo(absolute)
moveTo(relative)
```
but it is still not clear when each method should be invoked.

Reply:
May be there is some problem with the method naming

Reply:
I like the idea in general. What would you say if we don't split parameter signatures between move/non-moveTo, but just add a possibility to throw a runtime verification exception if withOffset(coordinates()) is provided  to moveTo and/or withOffset(offset())) is provided to non-moveTo method?

Reply:
@mykola-mokhnach I think that the good way is to inform users by method signature and compilation errors. I suppose users won't have fun if compiled will throw runtime verification errors.

The `Position` class has at least two purposes:
- unify signature of methods. Most of touch actions (present and further) can use either coordinate position/offset or offset from some element. So the class has 2 methods `withOffset` and `withElement` which invocation replaces previously stored value.

- bound used coordinate strategy (absolute or relative) by generic parameter. 

Currently most of actions use only points/elements, but I think there will be new actions which will use only  offsets/elements.

Reply:
If something breaks the spec then it should throw an error. However I see there is a major issue in the server implementation. iOS threats moveTo arguments as relative and Android as absolute.  Then it why such expression

```
       TouchAction dragNDrop = new TouchAction(driver)
                .longPress(longPressOptions()
                        .withPosition(coordinates(center1.x, center1.y))
                        .withDuration(ofSeconds(2)))
                .moveTo(position()
                        .withPosition(coordinates(center2.x, center2.y)))
                .release();
```

works perfectly for Android, but causes the unexpected behaviour for iOS

Reply:
@mykola-mokhnach Yes i remember i had an issue open on this at server end reg. `moveTo`

Reply:
Ok. I need to think a bit. 

One of the possible ways could be the ovelloading of such methods by iOS/Androind/Windows (if it will be added), maybe the making of TouchAction the abstract class with generic signature of such merhods, overloaded with bounds by sublasses .

Reply:
https://github.com/appium/appium/issues/7486